CorsixTH/Lua/app.lua

```diff
@@ -631,10 +630,23 @@ end
 
 --! Opens the given file name and returns all Level definitions in a table.
 --! Values in the returned table: "path", "level_file", "name", "map_file", "briefing", and "end_praise".
---!param level (string) Name of the file to read.
+--!param level (string) Full path or name of the file to read.
+--!param preferred_dir (string) Folder of the parent campaign file
 --!return (table) Level info found in the file.
-function App:readLevelFile(level)
-  local filename = self:getAbsolutePathToLevelFile(level)
+function App:readLevelFile(level, preferred_dir)
+  local filename
+  local search_paths = {
+    self.user_campaign_dir,
+    self.user_level_dir,
+    self.campaign_dir,
+    self.level_dir,
+  }
```

Comment:
Try to keep local variables as close to first use as possible.

It avoids accidentally using non-initialized variables, and makes it easier to find the declaration.

Reply:
Here, initializing `search_paths` just above 648 has the advantage that in some runs it never even creates the variable.

Reply:
Redone, with stricter error checking